### ISSUE TITLE
Over consumption of Heap memory after v0.32 (v.033 also)

### ISSUE THREAD

**Issue by opteemister** [DataDog/dd-trace-java#1012]
Hello. 

Pre-story:
We've upgraded dd-java-agent.jar to v.033 and started having problems with our services (java 8) running in a Docker container. All containers have limited memory. Services doesn't have limits for heap memory (no xms, xmx set). So in current configuration heap memory has dynamic size which depends on total memory available for a container. (e.g. container limit is 500mb, heap memory usually is about 125mb). 
Before dd-java-agent upgrade - heap memory usage was below the limit and GC didn't work much. But after upgrade we've started noticing that heap memory is not enough for service running:
1. GC major collection count was grown 10-20 times on service start (with same count of loaded class and thread count)
2. After start GC major collection continues for several services and sometimes service dies with OOM (not enough heap memory). 
3. After increasing total limit for container (heap memory limit also increases a bit) gc major collection was almost the same (high). And only with doubling memory (e.g. to 1G with heap memory 250Mb) gc major collection reduces to the normal level. But it is not acceptable to have 1GB reservation, because services doesn't need so much memory. 

After testing with different versions we've found that latest stable version (which doesn't have over consumption of heap memory) is 0.31.2. 
Affected versions are 0.32, 0.33

Here are several screenshots of heap memory, loaded classes, gc major collection metrics for services running with different versions of dd-java-agent.jar. Those metrics were recorded on service start. 

<img width="461" alt="Screenshot 2019-09-24 at 11 23 52" src="https://user-images.githubusercontent.com/13973153/65501267-5d806880-dec9-11e9-89b8-c20140c70906.png">
<img width="1250" alt="Screenshot 2019-09-24 at 11 25 32" src="https://user-images.githubusercontent.com/13973153/65501269-5d806880-dec9-11e9-8a6c-92d28bcd4f22.png">
<img width="906" alt="Screenshot 2019-09-24 at 11 29 11" src="https://user-images.githubusercontent.com/13973153/65501272-5d806880-dec9-11e9-9897-270ef47f17f3.png">

So we can't update to the latest version because our services dies with OOM. 
Could you please investigate why that happens and how we can avoid such behaviour?

**Comment by rutgermirabeau** [comment]
I tested the new 0.34.0 agent with a Java 11 Spring Boot application. Docker memory limit is 512mb with -XX:+UseContainerSupport -XX:MaxRAMPercentage=50 as memory settings in the JAVA_OPTS.

The 0.34.0 agent part in the graph is where the memory usage is almost 100%. With the 0.31.2 agent the memory usage is ~70% 
![Screenshot from 2019-10-15 13-17-19](https://user-images.githubusercontent.com/18399301/66827056-352add80-ef4e-11e9-8942-ce8130426cd0.png)

**Comment by tylerbenson** [comment]
@rutgermirabeau The problem you're seeing is we are trying to strike a balance between not using too much memory, but also keeping the startup time overhead minimal, all while avoiding adding an additional config setting.  Making that cache too small would result in discarding cached values that may need to be regenerated.  We tried to balance this in two ways.

1. The cache values are expired and cleared after a minute of no use.
2. The values are now stored as soft references.

Number 2 means that if the JVM is under memory pressure, it can discard values from the cache rather than throw an OOM error.

In summary, the greatest memory pressure from this cache is at startup time, when all the classes are being initially loaded, but after the application is running for a few minutes, the overhead should be reduced.

Do you have any suggestions for better ways to handle this?

**Comment by tylerbenson** [comment]
@opteemister please try out 0.34.0 and let us know if it helps.  I closed the issue for now, but we can reopen it if we need further consideration.

**Comment by opteemister** [comment]
@tylerbenson Thanks for the explanation. I'll try 0.34.0 and let you know the results.

**Comment by opteemister** [comment]
@tylerbenson I just upgraded to v0.35.0 and so far so good. Thank you for the help 👍
